server/utils/stigLibrary.ts

```diff
@@ -0,0 +1,291 @@
+            check.status = matchingCheck?.status;
+            check.comments = matchingCheck?.comments;
+            check.finding_details = matchingCheck?.finding_details;
+            check.severity_override = matchingCheck?.severity_override;
+            check.severity_justification = matchingCheck?.severity_justification;
+            check.AssessmentId = newAssessment.id;
+            await check.save();
+          }
+        }
+      }
+    }
+    return { results: "Success" };
+  } catch (error) {
+    logger.error(error);
+    throw error;
+  }
+};
+
+export const findMatch = (check: AssessmentItem, array: AssessmentItem[]) => {
+  // return array.find((item) => item.StigDatum?.vuln_num === check.StigDatum?.vuln_num);
+  return array.find((item) => item.StigDatum?.rule_id === check.StigDatum?.rule_id);
+};
+
+export const processLibrary = async (
+  sourceZip: string,
+  baseOutputPath: string,
+  originalName: string,
+): Promise<ProcessLibraryResults> => {
+  let outputDirPath = baseOutputPath;
+
+  if (originalName) {
+    outputDirPath = path.join(baseOutputPath, originalName.substring(0, originalName.length - 4));
+  }
+
+  if (!fs.existsSync(outputDirPath)) {
+    fs.mkdirSync(outputDirPath, { recursive: true });
+  }
+
+  const libraryNameAttributes = parseLibraryName(originalName);
+  const hash = await hashFile(sourceZip);
+
+  const newLibrary = await StigLibrary.build({
+    filename: libraryNameAttributes.filename,
+    hash,
+    classification: libraryNameAttributes.classification,
+    libraryDate: libraryNameAttributes.date,
+    version: libraryNameAttributes.version,
+    importedDate: DateTime.now().toISODate(),
+  });
+
+  try {
+    await newLibrary.save();
+  } catch (error) {
+    if (error instanceof UniqueConstraintError) {
+      error.errors.forEach((element) => {
+        logger.error(`[ERROR] ${libraryNameAttributes.filename} ${element.message}`);
+      });
+      fs.rmSync(sourceZip);
+      throw new Error("Error saving Library entry to DB.  Unique Constraint.", { cause: error });
+    } else {
+      logger.error("Error saving Library Entry.", { cause: error });
+      throw new Error("Error saving Library Entry.");
+    }
+  }
+
+  const newStigLibrary = await StigLibrary.findOne({
+    where: {
+      hash: newLibrary.dataValues.hash,
+    },
+  });
+
+  const filelist = await extractLibrary(sourceZip, outputDirPath);
+  fs.rename(sourceZip, path.join(outputDirPath, path.basename(sourceZip)), (error) => {
+    if (error) {
+      console.log(`[ERROR] Moving file ${error.message}`);
+    }
+  });
+
+  if (!newStigLibrary) {
+    logger.error("Unable to parse library version.");
+    throw new Error("Unable to parse library version.");
+  }
+
+  const processLibraryResults: ProcessLibraryResults = {
+    stigProcessed: 0,
+    newStigCount: 0,
+    updatedStigCount: 0,
+    unchangedStigCount: 0,
+    xmlsExtracted: filelist.length,
+  };
+
+  for (const xmlFile of filelist) {
```

Comment:
this feels like it's going to be one of the slowest parts of the ingest just cause we're forced to await on each stig finishing being processed before we can move on.  we discussed that the ingestion of these stigs needed to be done in order so as to prevent some conflicts that occasionally occurred.

i think we should spend some research time into seeing if we can as discussed do some pre-processing to figure out the dependency problem and then be able to go full async on upload into the db.